### ISSUE TITLE
Block I/O layer

### ISSUE THREAD

**Issue by liqinggd** [asterinas/asterinas#432]
# The Linux Storage Stack
![IO_stack_of_the_Linux_kernel](https://github.com/jinzhao-dev/jinux/assets/49150782/1e261bcf-e8d6-491c-99fc-788a579f1808)

# The core structures
Here we summarize the core structures used in the Linux kernel to accomplish the block I/O work. 

## Buffer Head
When a block is stored in memory(a buffer), each buffer is associated with exactly one block. The buffer serves as the object that represents a disk block in memory. Because the kernel requires some associated control information to accompany the data (such as from which block device and which specific block the buffer is), each buffer is associated with a descriptor. The descriptor is called a `buffer_head`.

The purpose of a `buffer_head` is to describe this mapping between the on-disk block and the physical in-memory buffer.
```c
struct buffer_head {
    unsigned long        b_state;          /* buffer state flags */
    struct buffer_head   *b_this_page;     /* buffers using this page */
    struct page          *b_page;          /* page storing this buffer */
    sector_t             b_blocknr;        /* logical block number */
    size_t               b_size;           /* block size (in bytes) */
    char                 *b_data;          /* pointer to data within the page */
    struct block_device  *b_bdev;          /* device where block resides */
    bh_end_io_t          *b_end_io;        /* I/O completion method */
    void                 *b_private;       /* data for completion method */
    struct list_head     b_assoc_buffers;  /* list of associated mappings */
    atomic_t             b_count;          /* buffer usage counter */
};
```

The physical block on the disk to which a given buffer corresponds is the `b_blocknr`-th logical block on the block device described by `b_bdev`.

The physical page in memory to which a given buffer corresponds is the page pointed to by `b_page`. More specifically, `b_data` is a pointer directly to the block (that exists somewhere in `b_page`), which is `b_size` bytes in length. Therefore, the block is located in memory starting at address `b_data` and ending at address (`b_data + b_size`).

The buffer head is a large and unwieldy data structure and it only describes a single buffer.

## Bio
The basic container for block I/O within the kernel is the `bio` structure. By allowing the buffers to be described in *segments*, the `bio` structure provides the capability for the kernel to perform block I/O operations of even a single buffer from multiple locations in memory (this is called scatter-gather I/O).
```c
struct bio {
        struct bio           *bi_next;          /* list of requests */
        struct block_device  *bi_bdev;          /* associated block device */
        blk_opf_t            bi_opf;            /* bottom bits REQ_OP, top bits req_flags. */
        unsigned long        bi_flags;          /* status and command flags */
        unsigned short       bi_ioprio;         /* io priority */
        blk_status_t         bi_status;         /* io status */
        struct bvec_iter     bi_iter;           /* io iterator */
        bio_end_io_t         *bi_end_io;        /* io completion method */
        void                 *bi_private;       /* owner-private method */
        unsigned short       bi_vcnt;           /* number of bio_vecs off */
        unsigned short       bi_max_vecs;       /* maximum bio_vecs possible */
        atomic_t             bi_cnt;            /* usage counter */
        struct bio_vec       *bi_io_vec;        /* bio_vec list */
};

struct bvec_iter {
        sector_t		bi_sector;	/* starting device address in 512 byte sectors */
        unsigned int		bi_size;	/* io count */
        unsigned int		bi_idx;		/* current index into bio_vec */
        unsigned int            bi_bvec_done;	/* number of bytes completed in current bvec */
};
```

In each given block I/O operation, there are `bi_vcnt` vectors in the `bio_vec` array starting with `bi_io_vec`. As the block I/O operation is carried out, the `bi_idx` field is used to point to the current index in the array. The relationship between them is shown below:
![image](https://github.com/jinzhao-dev/jinux/assets/49150782/095b6a53-8b22-47ef-8be1-ca70c6584202)

### The `bio_vec` structure
The `bi_io_vec` field points to an array of `bio_vec` structures. Segments are represented by `bio_vec`. These structures are used as lists of individual segments in this specific block I/O operation.

Each `bio_vec` is treated as a vector of the form `<page, offset, len>`, which describes a specific segment: the first physical page on which it resides, the location of the block as an offset into the page, and the length starting from the given offset. 
```c
struct bio_vec {
        struct page     *bv_page;  /* first physical page on which this segment resides */
        unsigned int    bv_len;    /* the length of this segment */
        unsigned int    bv_offset; /* the offset within the page where the segment resides */
};
```

### The role of `bi_idx` field
The `bi_idx` field is used to point to the current `bio_vec` in the list, which helps the block I/O layer keep track of partially completed block I/O operations.

A more important usage, however, is to allow the splitting of `bio` structures. With this feature, drivers such as RAID can take a single `bio` structure, initially intended for a single device, and split it up among the multiple hard drives in the RAID array. All the RAID driver needs to do is copy the `bio` structure and update the `bi_idx` field to point to where the individual drive should start its operation.

## The relationship between `buffer_head` and `bio`
The buffer represented by `buffer_head` acts as a single segment `bio`. Let's see the `submit_bh` method as an example:
```c
void submit_bh(blk_opf_t opf, struct buffer_head *bh)
{
	submit_bh_wbc(opf, bh, NULL);
}

static void submit_bh_wbc(blk_opf_t opf, struct buffer_head *bh,
			  struct writeback_control *wbc)
{
	const enum req_op op = opf & REQ_OP_MASK;
	struct bio *bio;

	if (buffer_meta(bh))
		opf |= REQ_META;
	if (buffer_prio(bh))
		opf |= REQ_PRIO;

	bio = bio_alloc(bh->b_bdev, 1, opf, GFP_NOIO);

	bio->bi_iter.bi_sector = bh->b_blocknr * (bh->b_size >> 9);

	__bio_add_page(bio, bh->b_page, bh->b_size, bh_offset(bh));

	bio->bi_end_io = end_bio_bh_io_sync;
	bio->bi_private = bh;

	/* Take care of bh's that straddle the end of the device */
	guard_bio_eod(bio);

	if (wbc) {
		wbc_init_bio(wbc, bio);
		wbc_account_cgroup_owner(wbc, bh->b_page, bh->b_size);
	}

	submit_bio(bio);
}
```

## Request
From the beginning of `submit_bio`, the `bio` enters the block I/O layer. These bios are abstracted into requests managed by the block I/O layer, and at the appropriate time, these requests leave the block I/O layer and enter the device driver layer.
```c
struct request {
	struct request_queue *q;
	struct blk_mq_ctx *mq_ctx;      /* for a software queue facing the submitting CPUs */
	struct blk_mq_hw_ctx *mq_hctx;  /* for a hardware queue facing the hardware block device */

	blk_opf_t cmd_flags;		/* op and common flags */
	req_flags_t rq_flags;

	int tag;
	int internal_tag;

	unsigned int timeout;

	/* the following two fields are internal, NEVER access directly */
	unsigned int __data_len;	/* total data len */
	sector_t __sector;		/* sector cursor */

	struct bio *bio;
	struct bio *biotail;

	union {
		struct list_head queuelist;
		struct request *rq_next;
	};

	struct block_device *part;
#ifdef CONFIG_BLK_RQ_ALLOC_TIME
	/* Time that the first bio started allocating this request. */
	u64 alloc_time_ns;
#endif
	/* Time that this request was allocated for this IO. */
	u64 start_time_ns;
	/* Time that I/O was submitted to the device. */
	u64 io_start_time_ns;

#ifdef CONFIG_BLK_WBT
	unsigned short wbt_flags;
#endif
	/*
	 * rq sectors used for blk stats. It has the same value
	 * with blk_rq_sectors(rq), except that it never be zeroed
	 * by completion.
	 */
	unsigned short stats_sectors;

	/*
	 * Number of scatter-gather DMA addr+len pairs after
	 * physical address coalescing is performed.
	 */
	unsigned short nr_phys_segments;

#ifdef CONFIG_BLK_DEV_INTEGRITY
	unsigned short nr_integrity_segments;
#endif

#ifdef CONFIG_BLK_INLINE_ENCRYPTION
	struct bio_crypt_ctx *crypt_ctx;
	struct blk_crypto_keyslot *crypt_keyslot;
#endif

	unsigned short ioprio;

	enum mq_rq_state state;
	atomic_t ref;

	unsigned long deadline;

	/*
	 * The hash is used inside the scheduler, and killed once the
	 * request reaches the dispatch list. The ipi_list is only used
	 * to queue the request for softirq completion, which is long
	 * after the request has been unhashed (and even removed from
	 * the dispatch list).
	 */
	union {
		struct hlist_node hash;	/* merge hash */
		struct llist_node ipi_list;
	};

	/*
	 * The rb_node is only used inside the io scheduler, requests
	 * are pruned when moved to the dispatch queue. special_vec must
	 * only be used if RQF_SPECIAL_PAYLOAD is set, and those cannot be
	 * insert into an IO scheduler.
	 */
	union {
		struct rb_node rb_node;	/* sort/lookup */
		struct bio_vec special_vec;
	};

	/*
	 * Three pointers are available for the IO schedulers, if they need
	 * more they have to dynamically allocate it.
	 */
	struct {
		struct io_cq		*icq;
		void			*priv[2];
	} elv;

	struct {
		unsigned int		seq;
		rq_end_io_fn		*saved_end_io;
	} flush;

	u64 fifo_time;

	/*
	 * completion callback.
	 */
	rq_end_io_fn *end_io;
	void *end_io_data;
};
```

## Request Queue
The design of the request queue went through two stages: single queue and multi-queue.

### Single-queue
Early storage devices were disks, characterized by mechanical movement addressing and lack of support for concurrent processing of I/O with multiple hardware queues. Therefore, the code logic naturally designed a software dispatch queue. This software logic architecture, with only one dispatch queue, is called a single-queue architecture.

![image](https://github.com/jinzhao-dev/jinux/assets/49150782/fe37d71c-f6a1-4550-a49f-ccfbf0fef42f)

### Multi-queue
When high-speed storage devices supporting multi-queues emerged, the processing time on the device side became shorter. The software overhead introduced by the single-queue approach became prominent, making the software a performance bottleneck. Therefore, a new multi-queue architecture was introduced. This architecture assigns a software queue to each CPU and distributes the same number of hardware context dispatch queues based on the number of hardware queues of the storage devices. This achieves concurrent processing and improves I/O performance.

This two-level design explicitly separates the two buffering functions of the staging area that was previously merged into a single queue in the Linux block layer: (i) support for IO scheduling (software level) and (ii) means to adjust the submission rate (hardware level) to prevent device buffer over run.

![image](https://github.com/jinzhao-dev/jinux/assets/49150782/ec172205-047b-46e1-b197-a53fa691e0da)

The blk-mq I/O Flow
![image](https://github.com/jinzhao-dev/jinux/assets/49150782/6f43f631-1f3e-4c7c-936a-dc30705fbbb1)

# The Draft Design
Here we choose to implement the muti-queue design. As we can see in the top picture, the `virtio-blk` supports multiple hardware queues. And the NVMe block devices are becoming increasingly popular.

```rust

/// The basic container for block I/O.
/// 
/// It contains two parts of information: the storage end and the memory end.
/// The storage end describes the blockdevice, the starting sector id and the length. 
/// The memory end contains several non-adjacent memory buffers(Segments),
/// allowing for the implementation of scatter-gather I/O.
struct Bio {
    /// The storage end
    storage_end: StorageEnd,
    /// The Memory end
    memory_end: MemoryEnd,
    /// The type, can be read/write/flush/discard/...
    type_: BioType,
    /// The bitflags
    flags: BioFlags,
    /// The state
    state: BioState,
    /// I/O completion method
    callback: BioCallback,
}

impl Bio {
    /// Construct a new bio.
    pub fn new(...) -> Self {
        todo!();
    }

    /// Add an I/O segment.
    pub fn add_segment(&mut self, segment: Segment) {
        todo!();
    }

    /// Submit the bio asynchronously.
    pub fn submit(&self) {
        let queue = self.block_device().request_queue();
        for request in queue.iter() {
            if request.insert(self) {
                return;
            }
        }
        let request = BioRequest::from(self);
        queue.push(request);
    }

    /// Submit the bio synchronously.
    pub fn submit_sync(&self) -> Result<()> {
        self.submit();
        // wait on the callback.
        // return the result
    }
}

/// The storage end part of bio
struct StorageEnd {
    /// The associated block device
    block_device: Arc<dyn BlockDevice>,
    /// The starting sector number for the block device.
    /// The smallest addressable unit on block device is a sector, generally 512 bytes.
    sector_id: Sid,
    /// The length
    len: usize,
}

/// The memory part of bio
struct MemoryEnd {
    /// Ensure the segments are not adjacent.
    /// If append a new segment, should try to merge into existed segment first.
    vec: Vec<Segment>,
}

/// Segment is a smallest unit in block I/O.
/// It is a contiguous memory area.
pub struct Segment {
    /// The contiguous page frames on which this segment resides.
    frames: VmFrameVec,
    /// The length, may cross pages.
    len: usize,
    // The offset relative to the first page.
    offset: usize,
}

impl Segment {
    /// Merge with another segment.
    ///
    /// Return it if can not merge.
    fn merge(&mut self, other: Segment) -> Option<Segment> {
        todo!();
    }
}

/// The block I/O request.
///
/// It ensures that the bio's type is same and the storage end is contiguous in sectors.
///
/// When submitting the bio, try to insert into an existing request if
/// the sector id is contiguous. Otherwise need to create a new request.
struct BioRequest {
    /// The Bios must be contiguous in sectors
    bios: Vec<Bio>,
    /// The first sector id
    sector_id: Sid,
    /// Total data length
    total_len: usize,
    /// The operation. can be OP_READ/OP_WRITE/OP_FLUSH/OP_DISCARD/...
    op: RqOperation,
    /// The flags
    flags: RqFlags,
    /// The timeout for this request
    timeout: Option<Duration>,
}

/// The two-level block I/O request queue.
///
/// It includes the software staging queues and the hardware dispatch queues.
///
/// The software staging queues map M:N to hardware dispatch queues.
/// There are always as many software queues as CPUs.
/// The number of hardware dispatch queues is determined by the driver's hardware queues. 
/// With enough hardware queues, it is a 1:1 mapping. 
struct RequestQueue {
    /// The per-core software staging queues.
    /// Handle the submissions from software.
    staging_queues: Vec<RwLock<VecDequeue<BioRequest>>>,
    /// The hardware dispatch queues.
    /// Handle dispatch to hardware and completions.
    dispatch_queues: Vec<RwLock<Ringbuf<BioRequest>>>
    /// Other information
    ...
}

/// Example of block device
pub struct BlockDevice {
    request_queue: RequestQueue,
}
```
